src/metafiles/v1.rs

```diff
@@ -42,6 +42,31 @@ impl CratesToml {
         fs::write(path, &toml::to_vec(&self)?)?;
         Ok(())
     }
+
+    pub fn append_to_path(
+        path: impl AsRef<Path>,
+        cvs: CrateVersionSource,
+        bins: BTreeSet<String>,
+    ) -> Result<(), CratesTomlParseError> {
+        let mut c1 = match Self::load_from_path(path.as_ref()) {
```

Comment:
@passcod I wonder is there anyway to perform this without race condition (e.g. use a lock file?) and without loading the existing file (which can be quite large)?

Reply:
Looking at the `.crates.toml`, I have some crazy thought:

If the file exists, we can serialize `CratesVersionSource` and `BTreeSet<String>`, then manually append to the file along with a newline.

If it is empty, then we just need to write `[v1]\n` to it.

Reply:
I don't think appends are particularly race safe.

Cargo does have a mechanism to lock the entire index, we could abuse that to effectively lock access to the `$CARGO_HOME`. I think we should go see other tools that write to these like `cargo-update` and `quickinstall` to see if they're doing anything about this. On supported platforms, we could also do advisory/file locking.

Optimisation-wise, with a JAX-style or evented or streaming parser we could have something that rewrites the files in one step rather than doing the load + modify + write, but I think that's a rather big change that would fit more in a standalone PR.

I also thing there is a bug in our handling at the moment. When we install a new version, this happens:

```

Reply:
I think we need to look into `cargo-install` to understand how to update these stuff properly, it sounds like we are doing it in the wrong way.

Reply:
@passcod Found it!

[`common_for_install_and_uninstall`](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/common_for_install_and_uninstall.rs) defines all the metadata for this.

Though I think it might be a good idea to create a format for our own, since the format that make sense to `cargo-install` but not `cargo-binstall`, e.g. `features`, `profile`, `rustc`, etc.

It is also not appendable, I will like to format to be appendable for installation though the `cargo-tools` can remove the duplicate entries when updating installed crates.

Co-op with `cargo-install` means we have to keep track of the upstream, otherwise we might accidentally break it

Reply:
Let's move this conversation to #176